### DISCUSSION TITLE
PowerShell Command Output Needs Work

### DISCUSSION THREAD

**Discussion by jdhitsolutions** [microsoft/winget-cli#2796] [PowerShell]
So I finally got the PowerShell module working. I know it is alpha, which is why I'm bringing this up now. Object output needs to be more PowerShell-like.  First glance, this looks good.
![image](https://user-images.githubusercontent.com/12089920/209220494-a2442631-0a5f-4473-a61f-071521316c60.png)

Until I try what most PowerShell users would try.

![image](https://user-images.githubusercontent.com/12089920/209220584-5c715611-1ba2-495c-8b64-5b7476f6f899.png)

Ok. I get it. You have a format file to handle output from complex object. I can work with that.

![image](https://user-images.githubusercontent.com/12089920/209220815-1e00f3ba-d382-41b6-9fa9-a827f7899c18.png)

I guess not. The NativeObject property is of no value to an IT Pro or end-user.

![image](https://user-images.githubusercontent.com/12089920/209221008-dc3ba022-4469-46c3-8eaa-21e9e4ab890e.png)

Command output should be a rich object, like system.diagnostics.process with the majority of properties "at the top". Your object is burying what the user wants to work with and makes this more complicated than it needs to be. Again, if Winget could return raw JSON, this would be so much easier than the COM object. I can't help but feel this approach is overkill.

**Comment by jdhitsolutions** [comment]
Here's another example of where the object fails.
![image](https://user-images.githubusercontent.com/12089920/209223508-13f7ece9-3613-4a4a-ad59-ff04eb1018f5.png)

and this:
![image](https://user-images.githubusercontent.com/12089920/209224104-14d9c58f-3dd7-4333-bc33-621097c8776b.png)

These break from expected PowerShell behaviors.

**Reply by whitTech** [reply]
The two examples provided here appear to be identical.

**Reply by jdhitsolutions** [reply]
Oops. I must have copied the wrong screen shot. I think this is what I meant to post. 
![image](https://user-images.githubusercontent.com/12089920/209240646-03cad13f-dce0-49d3-b20a-562df9bb9db2.png)

**Reply by whitTech** [reply]
Oh wow. No output is definitely a concerning level of broken. I might suggest editing the parent comment to include the corrected screenshot so it gets the visibility it deserves. 

Thanks for performing this important functionality validation, Jeff!

**Comment by jdhitsolutions** [comment]
This tells me you aren't writing a valid PowerShell object to the pipeline. 
![image](https://user-images.githubusercontent.com/12089920/209225059-9b4f3f70-da28-4cbe-9fbe-2358972adb3d.png)

**Comment by treestryder** [comment]
I'd rather see a valid Powershell .Net object, than JSON. Which could then be easily converted to JSON.
`Find-WinGetPackage terminal | ConvertTo-Json`

**Comment by jdhitsolutions** [comment]
I'm only suggesting JSON as a way to create a .NET object, even if it is a PSCustomObject. The problem is the convoluted nature of the COM object and there is different output from `Find-WingetGetPackage` and `Get-WingetPackage`. I would argue the object output should be identical. But even if it is different I expect to run a typical PowerShell command like:

` Find-WingetPackage git | Sort name | Select name,Version`

**Comment by jdhitsolutions** [comment]
I'd be happy with a traditional .NET object, It isn't here yet based on my testing.

**Comment by jdhitsolutions** [comment]
I think the Winget teams must answer a key design question: who is the target audience for the PowerShell module? I always tell my students, you need to consider who will be using your PowerShell commands? What are their expectations? What are the like use case scenarios you need to plan for? Answers to these questions help guide you on the choice of parameters, object property names, or whether you use parameter binding to name a few things.

**Reply by denelon** [reply]
The target audience is represented by the group of users who added their 👍 to:

<img width="284" alt="Who" src="https://user-images.githubusercontent.com/61799811/209234452-0ea6cf81-9a4a-44f0-9ebb-64d2950111fd.png">

We're excited to engage and see how users are wanting to use the product.

**Reply by jdhitsolutions** [reply]
If that is the case, which I assumed it was, then I think you have some major work to do. You have a lot of information that can be exposed as object properties, but they are buried in your COM object. You need to structure a PSCustomobject that brings those values to the surface. i can see where you might have one type of object for a Find command and a different object for a Get. I would suggest making the object as rich as possible because you don't know exactly what a user will need.

for example, in the Find result, I would start with these properties:
+ Name
+ ID
+ Version
+ Description
+ Publisher
+ PublisherHome
+ Source
+ Homepage
+ License
+ Tags
+ Author
+ ReleaseNotesURL
+ Copyright

The Get command should actually include the same information plus the installed version and if an update is available.

Then we can talk about default display properties and formatted views. You need to get the object output sorted out first.

**Reply by jdhitsolutions** [reply]
I do a lot of this with my module https://github.com/jdhitsolutions/WingetTools, but I have to parse text from winget output which is ugly, fragile, and next to impossible when other UI cultures come into play. 

BTW, there are 14K downloads of my module, so there is a bigger need than you realize.

**Reply by denelon** [reply]
I think there are going to be some subtle distinctions between metadata in a "manifest" and a "package". We will need to look through all of them to make the determination and figure out the right way to reason about things like "architecture" and "scope". The conceptual differences between what variations of a package are available vs. what is installed can be quite tricky to reason about, and in some cases with limited data, we aren't left with 100% confidence.

What did you use to generate the list of properties?

Why "PublisherHome" as opposed to "Publisher URL"?

**Reply by jdhitsolutions** [reply]
`PublisherURL`  is just as fine. That's the kind of thing where it is easy to add a property alias. But let't not get ahead of ourselves. And I have no problem in making a distinction between a manifest and a package. But know that many of the property names are identical and the distinction is likely to be lost on many users. PowerShell has a long history of abstracting the sausage-making. That's why you have to think about who is going to use your commands and how. Most users won't care. They are thinking about a "winget package" which may be abstracted from the manifest and package.  

Personally, I think you can make your life easier by focusing on a composite object. Otherwise, you need to make your commands very clear that these commands get manifest information about packages you might install, and these commands show you what is installed now.

**Reply by whitTech** [reply]
> The target audience is represented by the group of users who added their 👍 to:
> 
> <img alt="Who" width="284" src="https://user-images.githubusercontent.com/61799811/209234452-0ea6cf81-9a4a-44f0-9ebb-64d2950111fd.png">
> 
> We're excited to engage and see how users are wanting to use the product.

EDIT: Misunderstood the location of the engagement. Carry on.

In any case, I just want to thank you very much for your work thus far, @denelon. While I agree with Jeff that there remains a lot to do to make this a first class Powershell module, this is fantastic progress. Looking forward to using it (lots!).

**Reply by halr9000** [reply]
> I think the Winget teams must answer a key design question: who is the target audience for the PowerShell module?

Let's quote the manifesto for @denelon benefit:

[https://www.jsnover.com/blog/2011/10/01/monad-manifesto/]:

> **Problem**
> Windows has simple GUI administrative tools for basic users (Control Panel, MMC, etc). Windows also 
has a rich set of languages, APIs and object models for advanced systems programmers (C, C++, C#, 
WMI, Win32, .Net, etc). What is missing is the vital middle – administrator-oriented composable tools to 
type commands and automate management. The vital middle is typically addressed by scripting 
languages.

Also check out the visualization. 

In this case, the COM object and its interface is meant for systems programmers. Please don't expose that in powershell -- it misses the point and audience. Instead go higher level and orient to a task based workflow and interface. Give us something composeable.

**Comment by jantari** [comment]
EDIT: Ignore, I put what I originally posted here into a proper issue: https://github.com/microsoft/winget-cli/issues/2799

**Reply by denelon** [reply]
Yes, I would prefer smaller more discrete Issues. That makes it a bit easier to "complete" them within a single PR.

**Reply by jantari** [reply]
Totally makes sense, I filed https://github.com/microsoft/winget-cli/issues/2799

**Comment by treestryder** [comment]
I knew I recognized that name, @jdhitsolutions . It must be hard to teach [PowerShell Toolmaking in a Month of Lunches](https://www.manning.com/books/learn-powershell-toolmaking-in-a-month-of-lunches) through posts on GitHub issues. 😉

**Reply by jdhitsolutions** [reply]
Ha! I have been constantly teaching this stuff since 2006, seemingly everywhere. 😄

**Comment by treestryder** [comment]
Why was the package manager already built into Windows, the [PackageManagement](https://github.com/OneGet/oneget) Powershell provider, not used? Is WinGet that much better than OneGet? Was it forgotten? Do folks not realize Powershell is .Net, plus an amazing a command-line API?

**Reply by jdhitsolutions** [reply]
I think that module served a different need than winget. The PackageManagement module was focused on Nuget packages and I don't think winget has that same focus. I think there were other design considerations for winget early on which is why PowerShell support is only now getting attention.

**Reply by denelon** [reply]
TLDR;
There were environments that did not have PowerShell.

Discussed at:
* #221

**Reply by jdhitsolutions** [reply]
In PowerShell context, the package is a module when it comes to the PackageManagement module. 
![image](https://user-images.githubusercontent.com/12089920/209245632-7d8b00be-4900-4682-ac80-76b9f0f95e96.png)

**Reply by jdhitsolutions** [reply]
It doesn't matter now, but the idea that "there were environments without PowerShell" doesn't make sense. Winget, from what I can tell, only works on Windows machines and from a command prompt. Windows PowerShell has shipped with every version of Windows for years, especially desktops. It has been a while, but I suppose there might have been older versions of Windows Server that didn't ship with PowerShell out-of- the-box, but I can't imagine those systems were a target for winget.

And even if there was a requirement for Windows PowerShell 5.1, older systems should have been able to upgrade.  But this is all old news and I'm sure there have been changes to the Winget product team so the best we can do is soldier on from here.

**Comment by sba923** [comment]
Can't use the `CatalogPackage` "workaround" here:

![image](https://user-images.githubusercontent.com/12860484/212068930-a06117ed-d393-4298-bb5a-3db2481dacdf.png)

**Reply by sba923** [reply]
> Can't use the `CatalogPackage` "workaround" here:
> 
> ![image](https://user-images.githubusercontent.com/12860484/212068930-a06117ed-d393-4298-bb5a-3db2481dacdf.png)

@denelon I'd like to provide feedback on that first implementation, but without this `.CatalogPackage` trick I fear there's not much I can do...

**Reply by denelon** [reply]
@ryfu-msft has been working on improving the PowerShell implementation and fixing the bugs. I need to think about how we can tag issues specific to PowerShell (I did add a PowerShell tag) or the winget.exe CLI for issues that really only affect one or the other. We're going to work on releasing preview builds a bit more frequently as we work through the early PowerShell implementation.

**Reply by sba923** [reply]
Good. I'll probably wait 'til the next preview then.

**Reply by sba923** [reply]
> Good. I'll probably wait 'til the next preview then.

Problem's still there with v1.5.101-preview

![image](https://user-images.githubusercontent.com/12860484/212471296-9ef1490e-bae3-4aaf-99aa-1f4fd5523b02.png)

**Comment by denelon** [comment]
@ryfu-msft is working on the improvements to the objects returned.

* https://github.com/microsoft/winget-cli/pull/2871#event-8314801107

**Reply by sba923** [reply]
Great! Looking forward to the next preview with those improvements.

**Comment by sba923** [comment]
I'm experimenting with `1.5.441-preview`'s cmdlets: https://github.com/microsoft/winget-cli/releases/download/v1.5.441-preview/Microsoft.WinGet.Client-PSModule.zip.

I'm still lured by this:

![image](https://user-images.githubusercontent.com/12860484/224802819-86a3d454-3cf9-4b4b-87a9-5cbb81ac4fd9.png)

which seems to suggest one can do:

```powershell
get-wingetpackage | ? { $_.Available -ne '' }
```

@ryfu-msft suggests at https://github.com/microsoft/winget-cli/issues/3069#issuecomment-1466668531 to use this instead:

```powershell
get-wingetpackage | ? { $_.AvailableVersions -ne '' }
```

But this doesnt' seem to produce the expected results either:

![image](https://user-images.githubusercontent.com/12860484/224803609-b32a2ab4-be70-45b8-858f-7e8569c01c46.png)

Nor does

```powershell
get-wingetpackage | ? { $_.AvailableVersions -ne $null }
```

![image](https://user-images.githubusercontent.com/12860484/224803832-b0a4a7e8-ca78-48d0-89eb-f0d0ce4ed472.png)

**Reply by jantari** [reply]
Yea, the "Available" column is still very confusing. First you have to know that there is no object property called "Available" at all, it's just created during formatting. And then the value of "Available" is set like this:

```powershell
if ($_.IsUpdateAvailable) { $_.AvailableVersions[0] }
```

which is why it appears either empty when `IsUpdateAvailable` is false or as the latest available version when it is true because `AvailableVersions` is a sorted array of strings in which the first element is the latest available version.

To do what you want and show all packages for which an update is available, try:

```powershell
Get-WinGetPackage | Where-Object IsUpdateAvailable
```

Imo the "Available" column should either just be removed entirely (we can check `IsUpdateAvailable` and `AvailableVersions[0]` ourselves) or it should be introduced as a real property of type string. It would either contain the same as `AvailableVersions[0]` or be an empty string. Since empty strings evaluate to false in PowerShell we could still test against it like:

```powershell
# This code is hypothetical, doesn't work today
Get-WinGetPackage | Where-Object Available
```

It wouldn't be strictly neccessary but it would work intuitively and somewhat resemble the output of the DOS CLI.

**Reply by halr9000** [reply]
Yeah, ideally we could avoid violating the [principle of least surprise](http://principles-wiki.net/principles:principle_of_least_surprise) here by having the formatted output align to the property names returned by the object. This can be achieved a few ways. 

AliasProperty would help here! `Get-Process` & `System.Diagnostics.Process` is a great pattern that's existed since the beginning of time/powershell:

```
PS > ps | sort ws | select -last 5

 NPM(K)    PM(M)      WS(M)     CPU(s)      Id  SI ProcessName
 ------    -----      -----     ------      --  -- -----------
     92   600.01     288.36     170.48   16496   1 chrome
     28   451.29     344.76      26.78    6808   1 chrome
     75   373.59     394.10     346.11   17172   1 chrome
    374 …,219.02     662.60     108.59    3580   1 python
      0     5.21   1,810.10       0.00    2068   0 Memory Compression

PS > ps | gm work*,ws

   TypeName: System.Diagnostics.Process

Name         MemberType    Definition
----         ----------    ----------
WorkingSet   Property      int WorkingSet {get;}
WorkingSet64 Property      long WorkingSet64 {get;}
WS           AliasProperty WS = WorkingSet64
```

P.S. I would suggest not using space in column names during formatting. It's just gonna end in frustration every time. Nobody wants to type $.("Property Name"), even if it were a real property!

**Comment by sba923** [comment]
I know about the fact PowerShell does support displaying data in a way that's not a 1:1 match with the actual properties (I've been bitten e.g. by `Get-PSDrive`'s `Root` vs. `DisplayRoot`) but IMVHO that sort of works against the "easy discoverability" of properties.

True, one can learn from:

```powershell
PS> get-wingetpackage | gm

   TypeName: Microsoft.WinGet.Client.PSObjects.InstalledCatalogPackage

Name                  MemberType Definition
----                  ---------- ----------
CheckInstalledStatus  Method     string CheckInstalledStatus()
Equals                Method     bool Equals(System.Object obj)
GetHashCode           Method     int GetHashCode()
GetPackageVersionInfo Method     Microsoft.WinGet.Client.PSObjects.PackageVersionInfo GetPackageVersionInfo(string version)
GetType               Method     type GetType()
ToString              Method     string ToString()
AvailableVersions     Property   string[] AvailableVersions {get;}
Id                    Property   string Id {get;}
IsUpdateAvailable     Property   bool IsUpdateAvailable {get;}
Name                  Property   string Name {get;}
Source                Property   string Source {get;}
Version               Property   string Version {get;}
```

but for some reason that always comes _after_ I've hit some "unexpected behavior."

We need to find the right balance...
